Cupy test failures using eq_hist

Using current `master` (commit `a1de2d5e`) running `cupy` tests in `test_transfer_functions.py` there are test failures:

```
$ DATASHADER_TEST_GPU=1 pytest datashader/tests/test_transfer_functions.py
<snip>
datashader/tests/test_transfer_functions.py:503: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
datashader/transfer_functions/__init__.py:689: in shade
    return _interpolate(agg, cmap, how, alpha, span, min_alpha, name, rescale_discrete_levels)
datashader/transfer_functions/__init__.py:273: in _interpolate
    data = interpolater(data, mask)
datashader/transfer_functions/__init__.py:176: in eq_hist
    bin_centers = np.arange(vmin, vmax+1)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.

cupy/_core/core.pyx:1397: TypeError
========================================================================== short test summary info ===========================================================================
FAILED datashader/tests/test_transfer_functions.py::test_shade[None-a-agg1] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to constr...
FAILED datashader/tests/test_transfer_functions.py::test_shade_cmap_errors[agg1] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to c...
FAILED datashader/tests/test_transfer_functions.py::test_shade_rescale_discrete_levels[False-d-agg1] - TypeError: Implicit conversion to a NumPy array is not allowed. Plea...
FAILED datashader/tests/test_transfer_functions.py::test_shade_rescale_discrete_levels[True-d-agg1] - TypeError: Implicit conversion to a NumPy array is not allowed. Pleas...
================================================================== 4 failed, 107 passed, 6 skipped in 9.38s ==================================================================
 ```

This is because the `eq_hist` function uses some `numpy` calls that should be replaced with equivalent `cupy` calls if the supplied data is in a `cupy` array.

--- a/datashader/tests/test_transfer_functions.py
+++ b/datashader/tests/test_transfer_functions.py
@@ -169,6 +169,11 @@ def test_shade(agg, attr, span):
 
         img = tf.shade(x, cmap=cmap, how='eq_hist', rescale_discrete_levels=True)
         sol = tf.Image(eq_hist_sol_rescale_discrete_levels[attr], coords=coords, dims=dims)
+        if cupy and attr=='a' and isinstance(agg.a.data, cupy.ndarray):
+            # cupy eq_hist has slightly different numerics hence slightly different RGBA results
+            sol = sol.copy(deep=True)
+            sol[2, 0] = sol[2, 0] - 0x100
+
         assert_eq_xr(img, sol)
 
     img = tf.shade(x, cmap=cmap,

--- a/datashader/transfer_functions/__init__.py
+++ b/datashader/transfer_functions/__init__.py
@@ -160,28 +160,31 @@ def eq_hist(data, mask=None, nbins=256*256):
     """
     if cupy and isinstance(data, cupy.ndarray):
         from._cuda_utils import interp
+        array_module = cupy
     elif not isinstance(data, np.ndarray):
         raise TypeError("data must be an ndarray")
     else:
         interp = np.interp
+        array_module = np
 
     data2 = data if mask is None else data[~mask]
 
     # Run more accurate value counting if data is of boolean or integer type
     # and unique value array is smaller than nbins.
-    if data2.dtype == bool or (np.issubdtype(data2.dtype, np.integer) and data2.ptp() < nbins):
-        values, counts = np.unique(data2, return_counts=True)
-        vmin, vmax = values[0], values[-1]
+    if data2.dtype == bool or (array_module.issubdtype(data2.dtype, array_module.integer) and
+                               data2.ptp() < nbins):
+        values, counts = array_module.unique(data2, return_counts=True)
+        vmin, vmax = values[0].item(), values[-1].item()  # Convert from arrays to scalars.
         interval = vmax-vmin
-        bin_centers = np.arange(vmin, vmax+1)
-        hist = np.zeros(interval+1, dtype='uint64')
+        bin_centers = array_module.arange(vmin, vmax+1)
+        hist = array_module.zeros(interval+1, dtype='uint64')
         hist[values-vmin] = counts
         discrete_levels = len(values)
     else:
-        hist, bin_edges = np.histogram(data2, bins=nbins)
+        hist, bin_edges = array_module.histogram(data2, bins=nbins)
         bin_centers = (bin_edges[:-1] + bin_edges[1:]) / 2
         keep_mask = (hist > 0)
-        discrete_levels = np.count_nonzero(keep_mask)
+        discrete_levels = array_module.count_nonzero(keep_mask)
         if discrete_levels != len(hist):
             # Remove empty histogram bins.
             hist = hist[keep_mask]
@@ -189,7 +192,7 @@ def eq_hist(data, mask=None, nbins=256*256):
     cdf = hist.cumsum()
     cdf = cdf / float(cdf[-1])
     out = interp(data, bin_centers, cdf).reshape(data.shape)
-    return out if mask is None else np.where(mask, np.nan, out), discrete_levels
+    return out if mask is None else array_module.where(mask, array_module.nan, out), discrete_levels
 
 
 

